local_planner/src/nodes/waypoint_generator.cpp

```diff
@@ -339,13 +339,23 @@ void WaypointGenerator::nextSmoothYaw(float dt) {
 }
 
 void WaypointGenerator::adaptSpeed() {
-  speed_ = planner_info_.cruise_velocity;
+
+  //lowpass filter the speed to get smoother accelerations
+  float lp_filter_constant = 0.1;
+  float last_speed = speed_;
+
+  //at startup parameters are NAN (avoid propagating)
+  if (std::isfinite(planner_info_.cruise_velocity)) {
+    speed_ = planner_info_.cruise_velocity;
+  }else{
+    speed_ = 2.f;
+  }
 
   // If the goal is so close, that the speed-adapted way point would overreach
   float goal_dist = (goal_ - position_).norm();
   if (goal_dist < 1.f) {
-    output_.adapted_goto_position = goal_;
-
+    output_.goto_position = goal_;
+    speed_ = goal_dist;
```

Comment:
This could give you a step in the `speed_` when you hit this radius.

Reply:
the lowpass filter comes after that, so I think it should be fine, no?

Reply:
Ah, I misread the way your filter works. Typically you have one variable as the state, which gets updated gradually. Your filter instead 'clamps' your speed back to 90% of the previous speed. It works, it just isn't idiomatic.

It think it would be better to re-implement it clearly in terms of inputs and outputs, using the looprate-invariant formula from @bresch below.